### DISCUSSION TITLE
Proposal: Notifications API

### DISCUSSION THREAD

**Discussion by benpate** [bigskysoftware/htmx#1533] [Ideas]
A while back @1cg asked for suggestions on different Web APIs that HTMX/Hyperscript might simplify for developers.  This is a first pass at one of those APIs, and suggestions are actively requested for helping improve this idea.

The [Browser Notifications API](https://developer.mozilla.org/en-US/docs/Web/API/Notifications_API) uses Web Workers to access the global OS desktop notifications.  This seems like a great place for HTMX to help manage some complexity and provide some higher-level APIs.

## My Initial Proposal

* Register notification objects declaratively via HTMX attributes (`hx-notification=`?)
* Allow HTMX to use notification events as triggers
* Add listeners in WebSockets/SSE to trigger Notifications via some kind of mapping function
* Alternatively, we might decide that this is better to implement in Hyperscript.  

Notifications can accept a rich set of data, including various display properties, behaviors,etc..  I think we could define a simple, standard format for populating these object from JSON (or other) -encoded text.  This text would be included in the SSE/WS data and then populated into the Notification object.  Optionally, we might allow for a mapping function (`functionName` in my example code) for translating certain data packets into the required format.

### Server Sent Events

```html
<div hx-sse="connect:/my-event-source notify:eventName map:functionName">
```

### WebSockets
```html
<div hx-ws="connect:/my-websocket notify:eventName map:functionName">
```

**Comment by paxperscientiam** [comment]
I was thinking about this a bit earlier, and I definitely think that `htmx` could at least emit some sort of "notification" event.

It could look something like this:
1. Quad of new attributes: `htmx-notify` (generic), `htmx-notify-info`, `htmx-notify-warn`, and `htmx-notify-alert`
2. Implement associated events

Then using it is as simple as this:
```html
<!-- swap -->
<div htmx-notify-warn="This is a warning message">
  <!-- content -->
</div>
```
```js
htmx.onNotice(function (el, messge, status) { ... });
```
That said, not sure if this would be better suited to core or as an extension.

----

As a simplistic alternative for the interim, here's my pattern for delivering a message with a status.
1. Add data attributes to store message status and message value to parent element of swap.
```html
<div data-notify-status="some_value" data-notify-message="Some message">
  <!-- blah blah blah -->
</div>
```
2. Do something with values stored in said attributes upon, say, `htmx.onload` event for said element. In this example I'm using [notify.js](https://notifyjs.jpillora.com/)
```js
         htmx.onLoad(function(elt){
             const $elt = $(elt);
             const status = $elt.data('notify-status')
             const msg = $elt.data('notify-message');
             if (null != msg && null != status) {
                 $.notify(msg, status);
             }
         })
```

At any rate, I totally agree with @benpate 's proposal :)

**Comment by dz4k** [comment]
Instead of implementing notification support in htmx, a hypermedia-based notification library could be implemented:

 - htmx swaps response normally
 - response contains `<push-notification>` custom element
 - notification library picks it up and shows notification

This way, htmx and notifications are completely decoupled.

**Reply by benpate** [reply]
Yeah, this was just a brainstorm (3 years ago) when we were looking for other standard APIs that htmx might wrap. It is probably best as a separate library.. though I only see libraries with JS interfaces, not as `<custom-element>`s.
